perf/collector_libpfm.go

```diff
@@ -0,0 +1,250 @@
+}
+
+func NewCollector(cgroupPath string, events Events, numCores int) stats.Collector {
+	return &collector{cgroupPath: cgroupPath, events: events, cpuFiles: map[string]map[int]readerCloser{}, numCores: numCores}
+}
+
+func (c *collector) UpdateStats(stats *info.ContainerStats) error {
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+
+	stats.PerfStats = []info.PerfStat{}
+	klog.V(4).Infof("Attempting to update perf_event stats from cgroup %q", c.cgroupPath)
+	for name, files := range c.cpuFiles {
+		for cpu, file := range files {
+			buf := make([]byte, 32)
+			_, err := file.Read(buf)
+			if err != nil {
+				klog.Warningf("Unable to read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+				continue
+			}
+			perfData := &ReadFormat{}
+			reader := bytes.NewReader(buf)
+			err = binary.Read(reader, binary.LittleEndian, perfData)
+			if err != nil {
+				klog.Warningf("Unable to decode from binary format read from perf_event file (event: %q, CPU: %d) for %q", name, cpu, c.cgroupPath)
+			}
+			klog.V(4).Infof("Read metric for event %q for cpu %d from cgroup %q: %d", name, cpu, c.cgroupPath, perfData.Value)
+			scalingRatio := 1.0
+			if perfData.TimeEnabled != 0 {
+				scalingRatio = float64(perfData.TimeRunning) / float64(perfData.TimeEnabled)
+			}
+			stat := info.PerfStat{
+				Value:        uint64(float64(perfData.Value) / scalingRatio),
+				Name:         name,
+				ScalingRatio: scalingRatio,
+				Cpu:          cpu,
+			}
+			stats.PerfStats = append(stats.PerfStats, stat)
+		}
+	}
+	return nil
+}
+
+func (c *collector) Setup() {
+	cgroup, err := os.Open(c.cgroupPath)
+	if err != nil {
+		klog.Warningf("Unable to open cgroup directory %s: %s", c.cgroupPath, err)
+	}
+	defer cgroup.Close()
+
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+	cgroupFd := int(cgroup.Fd())
+	for _, group := range c.events.Events {
+		if len(group) == 1 {
+			customEvent, ok := c.events.eventToCustomEvent[group[0]]
+			if ok {
+				c.setupRawNonGrouped(customEvent, cgroupFd)
+			} else {
+				c.setupNonGrouped(string(group[0]), cgroupFd)
+			}
+		} else {
+			klog.Info("Grouped events are not supported yet")
+		}
+	}
+}
+
+func (c *collector) setupRawNonGrouped(event *CustomEvent, cgroup int) {
+	klog.V(4).Infof("Setting up non-grouped raw perf event %#v", event)
+	config := createPerfEventAttr(*event)
+	c.registerEvent(config, string(event.Name), cgroup)
+}
+
+func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, cgroup int) {
+	var cpu int
+	for cpu = 0; cpu < c.numCores; cpu++ {
+		pid, groupFd, flags := cgroup, -1, unix.PERF_FLAG_FD_CLOEXEC|unix.PERF_FLAG_PID_CGROUP
+		fd, err := unix.PerfEventOpen(config, pid, cpu, groupFd, flags)
+		if err != nil {
+			klog.Errorf("Setting up perf event %#v failed: %q", config, err)
+			return
+		}
+		perfFile := os.NewFile(uintptr(fd), name)
+		if perfFile == nil {
+			klog.Warningf("Unable to create os.File from file descriptor %#v", fd)
```

Comment:
Is there a reason why this is a warning rather than an error?  Is this expected in some cases?

Should we be returning these errors and logging them at the caller?

Reply:
If we consider this an error then we should probably `Destroy()` the collector as it will be in half-baked state. What do you think?

Reply:
Yeah.  I would prefer propagating errors from Setup() up to the manager, and having the manager initiate the teardown.  We can then log to notify that metrics for that container will not be available, like the GPU metrics does.

Reply:
I'm just looking at GPU implementation as I really want to keep them consistent. I will probably file another PR with some changes to interfaces defined in `stats` and their implementation in `accelerators`.

Reply:
ok, sgtm